polygon/heimdall/range_index.go

```diff
@@ -0,0 +1,93 @@
+package heimdall
+
+import (
+	"context"
+	"encoding/binary"
+
+	"github.com/c2h5oh/datasize"
+	"github.com/ledgerwatch/log/v3"
+
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon-lib/kv/mdbx"
+)
+
+type RangeIndex struct {
+	db kv.RwDB
+}
+
+const rangeIndexTableName = "Index"
+
+func NewRangeIndex(tmpDir string, logger log.Logger) (*RangeIndex, error) {
+	db, err := mdbx.NewMDBX(logger).
+		InMem(tmpDir).
+		WithTableCfg(func(_ kv.TableCfg) kv.TableCfg { return kv.TableCfg{rangeIndexTableName: {}} }).
+		MapSize(1 * datasize.GB).
+		Open(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	return &RangeIndex{db}, nil
+}
+
+func (i *RangeIndex) Close() {
+	i.db.Close()
+}
+
+func rangeIndexKey(blockNum uint64) [8]byte {
+	var key [8]byte
+	binary.BigEndian.PutUint64(key[:], blockNum)
+	return key
+}
+
+func rangeIndexValue(id uint64) [8]byte {
+	var value [8]byte
+	binary.BigEndian.PutUint64(value[:], id)
+	return value
+}
+
+func rangeIndexValueParse(value []byte) uint64 {
+	return binary.BigEndian.Uint64(value)
+}
+
+// Put a mapping from a range to an id.
+func (i *RangeIndex) Put(r ClosedRange, id uint64) error {
```

Comment:
FYI: if you plan high parallel RPS on this method - then can use `.Batch()` - to pay a bit of latency for high throughput.